README.md
Octave Toolbox for Computer-Generated Holograms (CGH)
=====================================================

Ulf GRIESMANN, 2020

A toolbox for [Octave](https://octave.org) to compute the layouts of
computer-generated holograms from scalar phase functions describing
the optical action of holograms in optical imaging systems. The
optical function of a hologram is generally modeled with optical
ray-tracing software and it can be encapsulated by a scalar optical
phase function $\phi$(x,y). The toolbox converts phase functions into
equivalent binary or multi-level holograms. The algorithm used in this
toolbox takes advantage of the relationship between the local
derivatives of phase functions and the local geometry (curvature) of
isophase lines. The isophase-following algorithm is easily extended
to phase functions with singularities and discontinuities that occur
in some optical applications. [A full description of the toolbox
algorithms has been published.](https://nvlpubs.nist.gov/nistpubs/jres/125/jres.125.024.pdf).
The [GDSII Toolbox](https://github.com/ulfgri/gdsii-toolbox) can be
used to create hologram layouts suitable for fabrication with
direct-write lithography systems. The CGH toolbox also
includes a family of functions for the robust and efficient estimation
and evaluation of Zernike polynomials, which are widely used in optical
applications.


Copyright
---------
The functions in this toolbox are in the **Public Domain**. For details
see the file LICENSE in the root directory of the toolbox.


Documentation
-------------
Additional documentation is available on:

https://sites.google.com/site/ulfgri/numerical/cgh-toolbox

in a user manual that describes the installation of the toolbox
and its use through a set of detailed examples.


Software Dependencies
---------------------
The CGH toolbox makes use of the [Octave Parallel Package](https://octave.sourceforge.io/parallel)
to speed up the computation of holograms on computers with multiple
CPU cores. The Parallel Package can now be used on all common
operating systems. In addition, the CGH toolbox requires the
[Octave Optimization Package](https://octave.sourceforge.io/optim). Symbolic
computations of Zernike polynomials require the
[Octave Symbolic Package](https://octave.sourceforge.io/symbolic) and an
installation of [Python](https://www.python.org) with
[SymPy](https://www.sympy.org/en/index.html).


Compiling
---------
The CGH toolbox contains several MEX functions that must be compiled
with a C compiler before the toolbox can be used.

For Octave on Linux, the mex functions are compiled by executing

$ ./makemex-octave

at the shell prompt. In Octave on Windows the mex functions are
compiled by changing to the ./cgh-toolbox directory and running

>> makemex

at the Octave command prompt.


Help
----
If you find a bug in the software, please send a message to
ulf.griesmann@nist.gov or ulfgri@gmail.com and I will try to correct it.


ChangeLog
14 October 2019, version 154
- use Frenet-Serret equation to calculate normal vector
  at predicted vertices

08 October 2019, version 151
- remove 'sorted' option from cgh parameters
- implemented an algorithm for selection of phase boundary
  start intersections that works with both continuous and
  dis-continuous phase functions.

06 October 2019, version 149
- remove test scripts from the toolbox directory tree

02 October 2019, version 146
- remove the option to specify tile-specific algorithms because it
  makes no sense.

25 September 2019, version 136,137
- changed the symmetric bracket search to an asymmetric
  search that searches for isophase brackets only in one 
  direction.
- removed 'phase2ps' because it's not ready for publication
- added 'polynomials' directory with functions for Zernike
  and disk harmonic function models.

22 September 2019, version 133
- fixed serious bug in phase2cgh/pa/isophase_clip.m Now works when
  the result of clipping is more than one polygon.

20 September 2019, version 132
- moved examples into own archive and prepare for testing

15 September 2019, version 123
- 'boundary_fill' extended to work with phase functions that
   have a phase singularity (Hilbert phase term)

12 September 2019, version 120
- completed phase boundary filling algorithm that only relies
  on edge intersections.

11 September 2019, version 119
- much improved 'hi' algorithm that does away with the concept
  of 'inner boundaries'. Now a straight-forward extension of
  the 'fi' algorithm - as it should be.

04 September 2019, version 110
- 'hi' algorithm for spiral zone plates with topological charge
  (Hilbert phase term) are beginning to work. The algorithm is 
  not really satisfying yet, though.

15 Aug 2019, version 80
- both pa and fi algorithms are finally working and seem robust.

26 July 2019, version 60
- completed the vectorized form of the isoline following algorithm.

21 Jul 2019, version 51
- rewrite from scratch the original implementation of the
  pilot approximation idea by J. Soons.

14 Oct 2018, version 40
- combine all CGH functions in one place
- subdirs for different algorithms
- 'phase2cgh' parameter structure has new flag to select algorithm

07 Oct 2018, version 39
- increased the increment h vor calculation of derivatives to 1.5e-2. A
  smaller value seems to result in rounding errors and numerical instability.
- implemented heuristic for tiles containing critical points

04 Oct 2018, version 37
- basic hologram layout calculation is all working, but needs thorough testing.

03 Oct 2018
- got 'phase2cgh_tile' working.

30 Sep 2018
- reached the "proof of concept" milestone. The isophase following algorithm is
  working, even though 'isophase_fill.m' is still buggy (boy was that true ... 09/2019).

26 Sep 2018
- finished coding all parts of 'phase2cgh_tile'.
- completed debugging of 'isophase_edge_xing.m' and 'isophase_edges.m'
  as well as a bevvy of auxiliary functions.


minterp2/minterp2.h
/* ANSI-C code produced by gperf version 3.1 */
/* Command-line: gperf minterp2.gperf  */
/* Computed positions: -k'1' */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
#error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gperf@gnu.org>."
#endif

#line 1 "minterp2.gperf"

/* hash table for option lookup
 *
 * process with:
 *    gperf minterp2.gperf > minterp2.h
 *
 * NOTE: this software is in the Public Domain
 * Ulf Griesmann, August 2013 
 */
#line 14 "minterp2.gperf"
struct keyword {
   char *name;
   double (*interp_func)(double*, double, double);
};

#define TOTAL_KEYWORDS 5
#define MIN_WORD_LENGTH 5
#define MAX_WORD_LENGTH 8
#define MIN_HASH_VALUE 5
#define MAX_HASH_VALUE 13
/* maximum key range = 9, duplicates = 0 */

#ifdef __GNUC__
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
hash (register const char *str, register size_t len)
{
  static unsigned char asso_values[] =
    {
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14,  5,
      14, 14, 14, 14, 14, 14, 14, 14,  5, 14,
      14, 14,  0, 14, 14,  0, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
      14, 14, 14, 14, 14, 14
    };
  return len + asso_values[(unsigned char)str[0]];
}

struct keyword *
in_word_set (register const char *str, register size_t len)
{
  static struct keyword wordlist[] =
    {
      {""}, {""}, {""}, {""}, {""},
#line 22 "minterp2.gperf"
      {"pchip",     &interp_pchip},
#line 23 "minterp2.gperf"
      {"spline",    &interp_pchip},
      {""}, {""}, {""},
#line 21 "minterp2.gperf"
      {"cubic",     &interp_cubic},
#line 20 "minterp2.gperf"
      {"linear",    &interp_linear},
      {""},
#line 24 "minterp2.gperf"
      {"lagrange",  &interp_lagrange}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register unsigned int key = hash (str, len);

      if (key <= MAX_HASH_VALUE)
        {
          register const char *s = wordlist[key].name;

          if (*str == *s && !strcmp (str + 1, s + 1))
            return &wordlist[key];
        }
    }
  return 0;
}
